torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+  //
+  // Also, since the inner most dim of different groups are not the same, it is
+  // natural to consider their vectorizability separately and allow them to have
+  // different vectorize/unroll sizes.
+
+  constexpr int64_t kSixteen = 16; // clang tidy
+
+  auto max_unroll_factor = ceilDiv(
+      // Available unrolling based on size of data type
+      (int64_t)kSixteen / max_input_dtype_size,
+      // Reduce max unrolling factor if we have many inputs/outputs to unroll
+      // as it could start consuming a lot of registers.
+      std::max(
+          (scheduler_utils::lastPow2(
+               (int64_t)grouped_inputs_outputs[0].size() +
+               (int64_t)grouped_inputs_outputs[1].size()) >>
+           2),
+          (int64_t)1));
+
+  // Don't unroll at the cost of getting a full wave on the GPU
+  auto max_unroll_factor_occupancy = ceilDiv(
+      n_elems,
+      device_multiprocessor_count * params.tile_size1 * params.tile_size2);
+  max_unroll_factor = std::min(max_unroll_factor, max_unroll_factor_occupancy);
+
+  // If we use RNG don't unroll so we can do correctness testing
+  if (fusion->isStochastic() && isDisabled(DisableOption::UnrollWithRng)) {
+    max_unroll_factor = 1;
+  }
+
+  // Compute maximum vectorize factor that can be used
+  size_t vectorize_factor1 = max_unroll_factor;
+  size_t vectorize_factor2 = max_unroll_factor;
+
+  for (auto tv : grouped_inputs_outputs[0]) {
+    const auto tv_vectorize_factor =
+        runtime_info.getInnerDimVectorizableWidth(tv);
+    vectorize_factor1 = std::min(vectorize_factor1, tv_vectorize_factor);
+  }
+  for (auto tv : grouped_inputs_outputs[1]) {
+    const auto tv_vectorize_factor =
+        runtime_info.getInnerDimVectorizableWidth(tv);
+    vectorize_factor2 = std::min(vectorize_factor2, tv_vectorize_factor);
+  }
+
+  // Try expanding vectorization to contig merged domains
+  auto expanded_vector_word_size1 =
+      scheduler_utils::expandVectorizationToContigMergedDomains(
+          fusion,
+          runtime_info,
+          grouped_inputs_outputs[0],
+          largest1,
+          0,
+          vectorize_factor1);
+  auto expanded_vector_word_size2 =
+      scheduler_utils::expandVectorizationToContigMergedDomains(
+          fusion,
+          runtime_info,
+          grouped_inputs_outputs[1],
+          largest2,
+          0,
+          vectorize_factor2);
+
+  expanded_vector_word_size1 = std::min(
+      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size1);
+  expanded_vector_word_size2 = std::min(
+      static_cast<size_t>(max_unroll_factor), expanded_vector_word_size2);
+
+  vectorize_factor1 = std::max(vectorize_factor1, expanded_vector_word_size1);
+  vectorize_factor2 = std::max(vectorize_factor2, expanded_vector_word_size2);
+
+  params.vectorize_factor1 = vectorize_factor1;
+  params.vectorize_factor2 = vectorize_factor2;
+
+  // TODO: should we adjust tile size according to max_unroll_factor?
```

Comment:
I'm more concerned about supporting tiles <32x32 in cases where we're latency bound. We haven't worried too much about vectorizing and unrolling, but could be potentially beneficial in some situations.

Reply:
This is definitely something we need to handle. Will do in followup PR